### PR TITLE
Improve UCX documentation and examples

### PR DESCRIPTION
**Pull request by charlesbluca**
Addressing #544, this PR aims to clarify the requirements, configuration, and usage of UCX with Dask-CUDA.

Still a lot to be done:

- [x] Flesh out hardware/software requirements
- [x] Rework CLI/Python usage examples
- [x] Clarify some uncertainties in the Configuration section
- [ ] Add standalone examples of UCX usage

### CODE REVIEW

**Comment by charlesbluca** [comment]
Thinking that the [UCX usage sections](https://dask-cuda.readthedocs.io/en/latest/ucx.html#launching-scheduler-workers-and-clients-separately) could probably be simplified a bit if we better fleshed out `dask-cuda-worker` and `dask_cuda.initialize(...)` in their own pages.

File: docs/source/ucx.rst
Comment (pentschev) [inline docs/source/ucx.rst:66]: I don't think all options under `ifconfig` are valid, for example `veth*` (often used by Docker) aren't. `ucx_info -d` only should be more reliable.

Also, interestingly the most important use case for us is missing in this section (it was missing before your proposed changes here too), which is the `"auto"` option, as documented in https://github.com/rapidsai/dask-cuda/blob/e72d776645ee3ee7e1672f73c9c043731cf401d5/dask_cuda/cli/dask_cuda_worker.py#L187-L203 and https://github.com/rapidsai/dask-cuda/blob/e72d776645ee3ee7e1672f73c9c043731cf401d5/dask_cuda/local_cuda_cluster.py#L72-L86 . Could you add that here as well? It's documented in "Important Notes" in this file below, but it seems sensible to have it here too.

File: docs/source/ucx.rst
Comment (pentschev) [inline docs/source/ucx.rst:77]: `net-devices="auto"` is an exclusive feature of Dask-CUDA, I don't think there are others that apply to this section. The line below this should also be clarifying in this context.

**Review by pentschev — COMMENTED** [review]
I added some comments here @charlesbluca , but it's looking good so far, thanks for working on this!

File: docs/source/ucx.rst
Comment (charlesbluca) [reply docs/source/ucx.rst:8]: Sure! More details would probably be better suited for a hardware requirements section.

File: docs/source/ucx.rst
Comment (charlesbluca) [reply docs/source/ucx.rst:25]: Thanks for the link! I will also go into some detail about when you would want to use the Conda package vs building/installing from source (essentially whether or not you plan to use IB/RDMA).

File: docs/source/ucx.rst
Comment (charlesbluca) [reply docs/source/ucx.rst:45]: Would it maybe be better to use that for the reasoning to enable it in this case?

File: docs/source/ucx.rst
Comment (charlesbluca) [reply docs/source/ucx.rst:59]: Is that to say that unexpected behavior will occur if `ucx.rdmacm` is enabled but `ucx.infiniband` is not?

File: docs/source/ucx.rst
Comment (charlesbluca) [reply docs/source/ucx.rst:66]: It does make sense to include information about `"auto"` here. Are there any resources on installing UCX-Py with hwloc support? Didn't see anything explicit about it in the installation guide.

File: docs/source/ucx.rst
Comment (charlesbluca) [reply docs/source/ucx.rst:77]: Yeah, so far from the config section alone I have:

- Use of CUDA transfers for task on GPU
- Setting `net-devices="auto"`

Anything I'm missing here?

File: docs/source/ucx.rst
Comment (pentschev) [reply docs/source/ucx.rst:25]: I think that information is better suited to the UCX-Py docs link I mentioned above. Let's try to avoid duplicating information which will eventually become inconsistent between the two docs.

File: docs/source/ucx.rst
Comment (pentschev) [reply docs/source/ucx.rst:45]: You can write a brief comment, but we probably don't want to rant about it here for too long as it's hard to draw a line of how much detail we want to go in. Here again a long rant would probably suit better UCX-Py (not suggesting we need to do it).

File: docs/source/ucx.rst
Comment (pentschev) [reply docs/source/ucx.rst:59]: Not unexpected behavior, the behavior is very much expected, resulting in an error on initialization.

File: docs/source/ucx.rst
Comment (pentschev) [reply docs/source/ucx.rst:66]: It's in the build dependencies: https://ucx-py.readthedocs.io/en/latest/install.html#build-dependencies

File: docs/source/ucx.rst
Comment (pentschev) [reply docs/source/ucx.rst:77]: > Use of CUDA transfers for task on GPU

This is again not true. A user can very much use mainline Dask with GPUs, they will only not benefit from the specializations Dask-CUDA provides -- thus my comment on the line that is immediately below this clarifying that difference. Therefore, the only thing I can think of here is indeed `net-devices="auto"`.

File: docs/source/ucx.rst
Comment (charlesbluca) [reply docs/source/ucx.rst:66]: Question about specifying an interface when InfiniBand is disabled:

I notice later on in the documentation, we say when setting up a scheduler for UCX that setting `DASK_UCX__NET_DEVICES` can generally be skipped when using a non-InfiniBand interface is used. Is this universally true, or does it only apply to the scheduler?

Comment (pentschev) [reply docs/source/ucx.rst:66]: This can apply to all processes. `UCX_NET_DEVICES` is merely a way to limit the visibility of network devices to UCX, meaning it won't "see" any others. We can still choose the device by specifying `--interface` to scheduler and workers, which is generally preferred if no IB is used. To be even more explicit, we're likely going to drop `DASK_UCX__NET_DEVICES` entirely starting from UCX 1.10, because UCX is now capable of ensuring it uses the IB interface topologically closest to the GPU (which is what `"auto"` does now).

**Comment by charlesbluca** [comment]
Thanks for the review @pentschev! Addressed them as best I could.

I reorganized the use cases into a general "Usage" section and consolidated most of the information on client setup to an individual "Client" section that showcases `dask_cuda.initialize`; feel free to let me know how that looks. My first thought is that it may be nested too deep when it gets to sections on setting up the scheduler + workers using the CLI.

**Comment by charlesbluca** [comment]
Is there anything we can do with the API reference to make `dask_cuda.initialize` show up with its arguments? It seems like it's meant for general use, but right now it seems like the only way to see what arguments it takes is to check the source code on GitHub.

**Comment by charlesbluca** [comment]
Made some changes to the API source so that `dask_cuda.initialize.initialize` should show up properly now; also moved the docstring for the `dask_cuda.initialize` module to the function itself and added some parameters.

File: dask_cuda/initialize.py
Comment (pentschev) [inline dask_cuda/initialize.py:44]: We need to add a note here now to let users know that this is already done for `LocalCUDACluster` (including the client, if it's instantiated on the same process as the cluster) and `dask-cuda-worker`, so users don't need to do that again. The common place where users would want to use this is when they instantiate their standalone client connecting to a cluster that has already been started with a combination of `dask-scheduler`/`dask-cuda-worker`s.

File: docs/source/ucx.rst
Comment (pentschev) [inline docs/source/ucx.rst:53]:
```suggestion
  This is recommended by UCX for use with InfiniBand, and will not work if InfiniBand is disabled.
```

File: docs/source/ucx.rst
Comment (pentschev) [inline docs/source/ucx.rst:73]:
```suggestion
    However, some features are exclusive to Dask-CUDA, such as the automatic detection of InfiniBand interfaces. 
```

File: docs/source/ucx.rst
Comment (pentschev) [inline docs/source/ucx.rst:88]: `LocalCUDACluster` is single-node only, anything else requires `dask-cuda-worker`.

File: docs/source/ucx.rst
Comment (pentschev) [inline docs/source/ucx.rst:108]: Yes, it is true.

File: docs/source/ucx.rst
Comment (pentschev) [inline docs/source/ucx.rst:128]: In most cases, no, UCX will be capable of identifying the optimal network interface to use.

File: docs/source/ucx.rst
Comment (pentschev) [inline docs/source/ucx.rst:194]: Yes, as in the previous comment in the scheduler section.

**Review by pentschev — COMMENTED** [review]
Thanks @charlesbluca , this is looking good so far!

I added a few minor suggestions and replied most of your questions (if I missed any, please let me know).

Since this is getting larger than I initially predicted, what do you say we merge this after my comments are addressed and continue with samples in a follow-up PR? I generally think this is better to avoid getting PRs too large and increasingly more difficult to track changes.

File: docs/source/ucx.rst
Comment (charlesbluca) [reply docs/source/ucx.rst]: In this case, should I remove the recommendation to explicitly set `ucx.net-devices` to an ethernet device when InfiniBand is disabled? Seems like an inconsistency, unless I'm missing something.

File: dask_cuda/initialize.py
Comment (charlesbluca) [reply dask_cuda/initialize.py:44]: Okay, in that case I will emphasize in the docstring that this script is meant for use with the client to mirror the pre-configured CUDA cluster/workers, and that the worker preload script would only need to be used with mainline Dask/Distributed.

**Comment by charlesbluca** [comment]
Addressed your comments - let me know if I should change anything. If not, I'm happy to merge and handle the examples in a separate PR! Thanks for the help 😄

File: docs/source/ucx.rst
Comment (pentschev) [reply docs/source/ucx.rst]: The way I read is we recommend users to set to the "optimal" device, so I would still leave it in place. I know in the past we had issues with suboptimal devices being chosen and it seems like UCX 1.10+ is better in that regard, so what I meant to say was that starting from UCX 1.10 we'll then prefer not to set `UCX_NET_DEVICES`, but I don't think in our current state (UCX 1.9), we're ready for that.

File: dask_cuda/initialize.py
Comment (pentschev) [inline dask_cuda/initialize.py:31]: But this function can't be used with `dask-scheduler`, so I don't think it really applies to that, only to the client. The scheduler will still have to rely on environment variables, so I would remove it from here.

**Review by pentschev — CHANGES_REQUESTED** [review]
One more small change, then I think we can merge it, but we're still blocked by https://github.com/rapidsai/dask-cuda/pull/546.

File: dask_cuda/initialize.py
Comment (charlesbluca) [reply dask_cuda/initialize.py:31]: If the scheduler still relies upon environment variables, does that mean that they must be set explicitly when running Python code?

I had initially removed the `os.environ[...]` calls from the `LocalCUDACluster` example as I thought all non-cluster/worker UCX config was handled through `initialize(...)` - I can add those back in if that's not the case.

Comment (pentschev) [reply dask_cuda/initialize.py:31]: We have 4 different types:

1. `LocalCUDACluster`: internally sets everything correctly to scheduler and workers, and also does that if `Client` is in the same process being instantiated after the cluster;
2. `dask-scheduler` (CLI): must set UCX options via environment variables (or `distributed.yaml`);
3. `dask-cuda-worker` (CLI): must  set UCX options via command-line arguments;
4. `Client` (standalone process): must set UCX options via environment variables (I think `distributed.yaml` also works here) or may use `dask_cuda.initialize`.

So what I meant was that for case 2 we can't use `dask_cuda.initialize` because it's a standalone process started via CLI. On the other hand, for `Client` we can either choose to use `dask_cuda.initialize` or environment variables, thus I think it only makes sense documenting for use with the client, so I would change the sentence above to:

"This is especially important for the ~scheduler and~ client, ..."

**Comment by jakirkham** [comment]
rerun tests

**Review by pentschev — APPROVED** [review]
LGTM, thanks @charlesbluca for the work on improving docs here!

**Comment by pentschev** [comment]
@gpucibot merge

**Pull request merged**
